### PR TITLE
Fixes for HDA-DMA driver

### PR DESCRIPTION
**Pull request by ranj063**
Part 1 of HDA DMA fixes. Fixes the handling of FIFORDY bits and pause/release cleanups for host/link DMA

### CODE REVIEW

File: src/drivers/intel/hda/hda-dma.c [hunk 1]
```diff
@@ -756,9 +759,9 @@ static int hda_dma_set_config(struct dma_chan_data *channel,
 	     config->src_width <= 2))
 		dgcs |= DGCS_SCS;
 
-	/* set DGCS.FIFORDY for output dma */
-	if ((config->cyclic && config->direction == DMA_DIR_MEM_TO_DEV) ||
-	    (!config->cyclic && config->direction == DMA_DIR_LMEM_TO_HMEM))
+	/* set DGCS.FIFORDY for host output dma */
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:731]: the comment isn't aligned with the code now, we deal with both directions.

File: src/drivers/intel/hda/hda-dma.c [hunk 2]
```diff
@@ -756,9 +759,9 @@ static int hda_dma_set_config(struct dma_chan_data *channel,
...
+	if ((config->cyclic && config->direction == DMA_DIR_HMEM_TO_LMEM) ||
+	    (config->cyclic && config->direction == DMA_DIR_LMEM_TO_HMEM))
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:733]: I think the !config->cyclic should be treated as a separate patch since that part was always false.

File: src/drivers/intel/hda/hda-dma.c [hunk 3]
```diff
@@ -407,8 +407,7 @@ static int hda_dma_enable_unlock(struct dma_chan_data *channel)
 	}
 
 	/* start link output transfer now */
-	if (channel->direction == DMA_DIR_MEM_TO_DEV &&
-	    !(hda_chan->state & HDA_STATE_RELEASE))
+	if (channel->direction == DMA_DIR_MEM_TO_DEV)
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:407]: I wasn't able to follow that simplification.

File: src/drivers/intel/hda/hda-dma.c [hunk 4]
```diff
@@ -570,10 +569,6 @@ static int hda_dma_release(struct dma_chan_data *channel)
 	 */
 	hda_chan->state |= HDA_STATE_RELEASE;
 
-	if (channel->direction == DMA_DIR_HMEM_TO_LMEM ||
-	    channel->direction == DMA_DIR_LMEM_TO_HMEM)
-		ret = hda_dma_host_start(channel);
-
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:577]: Same, not able to follow this simplification - not sure why we did a host_start in a dma_release in the first place.

File: src/drivers/intel/hda/hda-dma.c [hunk 5]
```diff
@@ -829,12 +829,16 @@ static int hda_dma_link_check_xrun(struct dma_chan_data *chan)
...
 		tr_err(&hdma_tr, "hda_dma_link_check_xrun(): underrun detected");
 		dma_chan_reg_update_bits(chan, DGCS, DGCS_BUR, DGCS_BUR);
-	} else if (chan->direction == DMA_DIR_DEV_TO_MEM && dgcs & DGCS_BOR) {
+		return -ENODATA;
+	}
+
+	if (chan->direction == DMA_DIR_DEV_TO_MEM && (dgcs & DGCS_BOR)) {
 		tr_err(&hdma_tr, "hda_dma_link_check_xrun(): overrun detected");
 		dma_chan_reg_update_bits(chan, DGCS, DGCS_BOR, DGCS_BOR);
+		return -ENODATA;
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:841]: doesn't this need additional changes in the caller as mentioned by @lyakh ?

**Review by plbossart — COMMENTED** [review]
Not fully clear on some simplifications - but the initial code is already confusing...

File: src/drivers/intel/hda/hda-dma.c (hunk 5 above)
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:841]: I think his comment was for my host changes and I fixed that in my other PR. This one is correct isnt it?

File: src/drivers/intel/hda/hda-dma.c (hunk 3 above)
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:407]: the only thing release ever did was to start host DMA but we never pause/release the host DMA in the first place, so the preovious patch removed that. With that, release literally does nothing now

Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:407]: but we still set the HDA_STATE_RELEASE bit? I am officially confused!

File: src/drivers/intel/hda/hda-dma.c (hunk 5 above)
Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:841]: I can't recall precisely the issue... Just wanted to check that this change is legit.

File: src/drivers/intel/hda/hda-dma.c (hunk 3 above)
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:407]: that bit is used to do 2 things: 1. releae host DMA which will never happen because we do no release host DMA. 2. prevent incrementing the firmware pointer for release which actually results in an xrun. But it isnt a register BIT just a state bit we used for the above 2 things. So its no longer needed

Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:407]: so can we remove line 414 as well and even the definition of HDA_STATE_RELEASE?

Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:407]: but I have removed the defitnion of HDA_STATE_RELEASE, no? not sure what 414 is, probably clearing the state and I' have removed that as well

**Comment by fredoh9** [comment]
SOFCI TEST

File: src/drivers/intel/hda/hda-dma.c (hunk 4 above)
Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:577]: yeah, this change looks unrelated to RELEASE removal ? Line 570 also sets the RELEASE state, should we remove RELEASE completely ?

File: src/drivers/intel/hda/hda-dma.c (hunk 3 above)
Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:407]: TBH, I would lump all the HDA DMA driver RELEASE removal into one patch.

Comment (lyakh) [inline src/drivers/intel/hda/hda-dma.c:407]: I'm probably missing something, I thought `DMA_DIR_MEM_TO_DEV` was only for link DMA playback?

**Review by lgirdwood — COMMENTED** [review]
Quite a lot of red on CI pause tests, this PR could be split to upstream the low risk updates first.

**Review by lyakh — CHANGES_REQUESTED** [review]

**Comment by ranj063** [comment]
@plbossart @lgirdwood I've dropped a couple of patches in this series. 1. to drop the release for link DMA. Even though this helps with the xrun, it doesnt completely avoid it. 2. return the right error in case of xrun. This is still TBD. We'll have to wait until I have a proper working solution for this.

File: src/drivers/intel/hda/hda-dma.c [hunk 6]
```diff
@@ -571,10 +570,6 @@ static int hda_dma_release(struct dma_chan_data *channel)
 	 */
 	hda_chan->state |= HDA_STATE_RELEASE;
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:571]: is the comment and assignment still valid, if you don't have a work cb any longer?

Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:571]: the assignment is valid because it helps avoid the FP increment during start for the release case. I'll remove the comment.

Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:567]: what does 'FP' stand for @ranj063, I wasn't able to follow your answer.

Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:567]: firmware pointer increment. we seem to increment it in the case of start and avoid it during release. I am not sure why though. I left it in as is for now as it doesnt really help avoid the xrun issue completely

**Comment by lgirdwood** [comment]
@wszypelt @kkarask looks like CI has stalled - can you unblock it ? Thanks !

**Comment by marc-hb** [comment]
https://sof-ci.01.org/sof-pr-viewer/#/build/PR4820/build7404529 is green but incomplete.

**Comment by marc-hb** [comment]
https://sof-ci.01.org/sofpr/PR4820/build10517/devicetest/ is green but missing all SOF logs because of some internal issue. Restarting it.

**Comment by marc-hb** [comment]
SOFCI TEST

File: src/drivers/intel/hda/hda-dma.c [hunk 7]
```diff
@@ -756,8 +759,8 @@ static int hda_dma_set_config(struct dma_chan_data *channel,
 	     config->src_width <= 2))
 		dgcs |= DGCS_SCS;
 
-	/* set DGCS.FIFORDY for output dma */
-	if (config->direction == DMA_DIR_MEM_TO_DEV || config->direction == DMA_DIR_LMEM_TO_HMEM)
+	/* set DGCS.FIFORDY for host DMA for both playback and capture */
+	if (config->direction == DMA_DIR_HMEM_TO_LMEM || config->direction == DMA_DIR_LMEM_TO_HMEM)
 		dgcs |= DGCS_FIFORDY;
 
 	dma_chan_reg_write(channel, DGCS, dgcs);
```
Comment (keyonjie) [inline src/drivers/intel/hda/hda-dma.c:757]: Since we set/clear DGCS_FIFORDY explicitly in start/stop, do we need to preset it in _set_config() here?

File: src/drivers/intel/hda/hda-dma.c [hunk 8]
```diff
@@ -348,9 +348,8 @@ static int hda_dma_host_start(struct dma_chan_data *channel)
 	struct hda_chan_data *hda_chan = dma_chan_get_data(channel);
 	int ret = 0;
 
-	/* Force Host DMA to exit L1 only on start*/
-	if (!(hda_chan->state & HDA_STATE_RELEASE))
-		pm_runtime_put(PM_RUNTIME_HOST_DMA_L1, 0);
+	/* Force Host DMA to exit L1 on start */
+	pm_runtime_put(PM_RUNTIME_HOST_DMA_L1, 0);
```
Comment (keyonjie) [inline src/drivers/intel/hda/hda-dma.c:352]: I think the original logic is that we don't need to force L1 exit for the pause/release iteration, while we need it for a _start, as there will be preloading for a hda_dma_start() while there isn't for the _release().

We do have dma_pause() and dma_release() calling in dai.c, even though there might not be trigger pause/release IPCs from the Linux host side, so I would suggest to keep this as it is in case it doesn't bring any issues like Xruns to us.

File: src/drivers/intel/hda/hda-dma.c [hunk 9]
```diff
@@ -979,8 +979,6 @@ const struct dma_ops hda_host_dma_ops = {
 	.start			= hda_dma_start,
 	.stop			= hda_dma_stop,
 	.copy			= hda_dma_host_copy,
-	.pause			= hda_dma_pause,
-	.release		= hda_dma_release,
```
Comment (keyonjie) [inline src/drivers/intel/hda/hda-dma.c:990]: "hda_dma_pause/release declared but not used"?

File: src/drivers/intel/hda/hda-dma.c (hunk 8 above)
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:352]: @keyonjie DAI uses link DMA and we have the ops for link DMA here. There is no need for host DMA pause/release when it is not used

File: src/drivers/intel/hda/hda-dma.c (hunk 9 above)
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:990]: sorry dont get it? it is used in link DMA ops

Comment (keyonjie) [reply src/drivers/intel/hda/hda-dma.c:990]: > sorry dont get it? it is used in link DMA ops

Oh, makes sense then, maybe better to rename them to hda_dma_link_pause/release().

File: src/drivers/intel/hda/hda-dma.c (hunk 7 above)
Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:757]: this seems to directly contradict the stated purpose of f96878a8f4e50e5b8023ddcf8810ea211b1899c3

**Comment by lgirdwood** [comment]
@ranj063 CI looks good around the pause states :)  are all the opens resolved now ?

**Comment by ranj063** [comment]
> @ranj063 CI looks good around the pause states :) are all the opens resolved now ?

@lgirdwood I dont have any opens. As for @lyakh's comment, the check for FIFORDY bit in set_cnofig is inconsequential anyway, we were setting it unconditionally during START anyway. And I confirmed that we should set it for host DMA for playback and capture

File: src/drivers/intel/hda/hda-dma.c (hunk 7 above)
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:757]: @lyakh I'm not sure what that specific commit achieved anyway. Before my change, we end up setting the FIFORDY bit for all DMAs unconditionally during START anyway.

**Comment by marc-hb** [comment]
SOFCI TEST

File: src/drivers/intel/hda/hda-dma.c (hunk 7 above)
Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:757]: @ranj063 not sure GH made it clear enough what my comment was referring to. I meant the removal of cyclic support as per your commit comment "config->cyclic is not relevant for HDA DMAs." I'm not sure whether your comment is correct, but I just commented, that it contradicted the original commit, that introduced cyclic DMA support for HDA. Don't know whether that commit was right though.

Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:757]: Maybe best to state in the comments the current feedback on setting ```DGCS.FIFORDY``` from KarL. This way everyone can understand.

Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:757]: @lgirdwood I have updated the commit message and added comments in the next patch for FIFORDY bit updates.

**Review by plbossart — APPROVED** [review]
I only have one comment, but that could be handled in a follow-up PR

File: src/drivers/intel/hda/hda-dma.c (hunk 7 above)
Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:758]: I think I was confused by the commit description phrase "`config->cyclic` is not relevant for HDA DMAs." It probably means "this one specific use of `config->cyclic` is irrelevant." If it really is trying to say, that the whole `config->cyclic` support in HDA DMA is irrelevant, then probably the whole commit https://github.com/thesofproject/sof/commit/f96878a8f4e50e5b8023ddcf8810ea211b1899c3 should be reverted.

**Review by lgirdwood — APPROVED** [review]
